### PR TITLE
Added CRD and design documents

### PR DESCRIPTION
**Pull request by shtripat**
monstorak-issue-id: monstorak/monstorak/issues/6
monstorak-issue-id: monstorak/monstorak/issues/22
Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

**Comment by shtripat** [comment]
@aruniiird @umangachapagain please review

**Comment by umangachapagain** [comment]
monstorak/pull/25 Captures my thought process. We could combine both.

**Comment by shtripat** [comment]
> ures my thought process. We could combine both.

@umangachapagain if you like, these changes can be merged with /pull/25 and I would abandon this PR.

**Review by aruniiird — COMMENTED** [review]
Making it inline with Umanga's PR#25

**Comment by shtripat** [comment]
@umangachapagain I added the steps for deployment of operator. Still one thing I am not clear is that, where are we taking care of enabling monitoring for new namespace. Can you please have a look comment for required changes in the PR?

@cloudbehl ^^^

File: docs/Users_guide/index.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# User documentation
+
+For now, there's not much to use. Check back later when we have something you
```
Comment (umangachapagain) [inline docs/Users_guide/index.md]: We do have something to use.

File: docs/Developers/Design/deploy_procedure_design.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+corresponding entry populated in
+`.Status.ReconcileActions map[string]reconciler.Result`.
+
+Top level actions are executed in an arbitrary order so they must define any
+prerequisite actions explicitly.
+
+An action may be a top-level action and still defined as a prerequisite and the
+caching implementation will ensure that it is executed a maximum of once per
+Procedure execution.
+
```
Comment (umangachapagain) [inline docs/Developers/Design/deploy_procedure_design.md:3]: Line 3 to 13 makes no sense here.

File: docs/Developers/Design/deploy_procedure_design.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+![ActionGraph](workflow.dot.png)
+
+# StorageMonitoringPlan actions
```
Comment (umangachapagain) [inline docs/Developers/Design/deploy_procedure_design.md]: This title is not so meaningful here. Should be  `Prerequisites` and each sub-topic (i.e. left half of the diagram) should be explained.

File: docs/Developers/Design/deploy_procedure_design.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+## clusterMonitoringDeployed
+
+## storageClusterDeployed
+
+## monitoringEnabledForStorage
+
+## deployStorageMixin
```
Comment (umangachapagain) [inline docs/Developers/Design/deploy_procedure_design.md]: This is not a sub heading.

File: docs/Developers/Design/deploy_procedure_design.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+The monstorak operator consists of two controllers. The first one takes care of
```
Comment (umangachapagain) [inline docs/Developers/Design/deploy_procedure_design.md]: Needs a heading. Maybe `Deploy Storage Mixin`. And content should be more about mixin.

File: docs/Developers/Design/deploy_procedure_design.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+getting the instanciating/reconciating customer resources like StorageCatalog,
```
Comment (umangachapagain) [inline docs/Developers/Design/deploy_procedure_design.md]: `instanciating/reconciating` is not a word. Reform the sentence to use the `instantiation` and `reconciliation`

File: docs/Developers/Design/deploy_procedure_design.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+StorageAlert and StorageDasboard. Once customer resources are available, the
+second controller forms storage monitoring plans and executes them one by one
+based on how many storage system and what artificats to be deployed.
+
+# Deploy Operator with mixins
```
Comment (umangachapagain) [inline docs/Developers/Design/deploy_procedure_design.md]: `Deploy Operator with mixins` is wrong. Should be `Deploy mixins with Operator` and should explain how.

File: docs/Developers/Design/deploy_procedure_design.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+To deploy the operator simply deploy to your cluster. The set of steps to be followed
+for deployment of monstorak operator and related artifacts, follow the below steps
+
+## Create the namespace `storage-monitoring`
```
Comment (umangachapagain) [inline docs/Developers/Design/deploy_procedure_design.md]: These steps are not specific to design. Should be in the main README file in the project root.

File: docs/Developers/Design/deploy_procedure_design.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+```
+
+## Deploy the custom resources
+```
+oc create -f deploy/kubernets/05_storageAlerts.yaml
+```
+
+## Deploy the storage alert rules
+```
+oc create -f jsonnet/manifests/ceph-prometheus-rules.yaml
```
Comment (umangachapagain) [inline docs/Developers/Design/deploy_procedure_design.md]: Why do we need this? If we do this, we don't need an operator at all. This step is what the operator will perform internally based on the CRs created.

File: docs/Developers/Design/CRD_design.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+# Overview
+
+A single Monstorak operator can take care of deployment of mixins for one or
+more storage systems. Each storage cluster should already be deployed within
+the kubernetes cluster, or the storage cluster could be running on hosts outside
+of kubernetes cluster. The capabilities of the operator remains same in both
+the modes of the deployment of storage cluster and the same set of CRDs should
+be used for both the cases.
+
+A given deployment configuration for storage systems contains a Custome Resource
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md]: `Custome` -> `Custom`

File: docs/Developers/Design/CRD_design.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+which contains configurations options like whether alerts are enabled, if
+dashboards are enabled, name of the storage system and version details of storage
+system.
+
+![Hierarchy of monstorak customer resources](crd_hierarchy.dot.png)
+
+# Custom Resources Definitions
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md]: `Resources` -> `Resource`

File: docs/Developers/Design/CRD_design.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+This section describes the fields in each of the custom resources.
+
+## **StorageCatelog**
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md:30]: Spellcheck

File: docs/Developers/Design/CRD_design.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+### *Singular: StorageCatalog, Plural:StorageCatalogs*
+
+The StorageCatalog custom resoyrce defines a storage system for which mixins could
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md]: spellcheck

File: docs/Developers/Design/CRD_design.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+kind: StorageAlert
+spec:
+  - storage: "ceph"
+    - version: "mimic"
+      alert: false
+    - version: "nautilus"
+      alert: true
+      prometheus:
+      - namespace: "monitoring"
+      - lable: "prometheus:k8s"
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md]: `lable` -> `label`

File: docs/Developers/Design/CRD_design.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+```
+
+## **StorageDashboard**
+
+### *Singular: StorageDashboard, Plural: StorageDashboards*
+
+The StorageDashboard customer resource defines the dashboards configuration for
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md]: `customer` -> `custom`

File: docs/Developers/Design/CRD_design.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+a storage system. This basically captures details like for which version
+dashboards are enabled, if dashboards enabled then which namepsace the
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md:78]: spellcheck

File: docs/Developers/Design/CRD_design.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+      grafana:
+      - namespace: "monitoring"
+      - label: "grafana:k8s"
+```
+
+## **StorageMonitoringPlan**
+
+### *Singular: StorageMonitoringPlan, Plural: StorageMonitoringPlans*
+
+The StorageMonitoringPlan custom resource defines a plan which shoul be
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md]: spellcheck

File: docs/Developers/Design/CRD_design.md [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+  - storage: "ceph"
+  - version: "mimic"
+    - alert: false
+    - dashbaord: false
+  - version: "nautilus"
+    - alert: true
+    - dashboard: true
+```
+
+All CRs live within the `mixins.monstorak.org` group and have version
```
Comment (umangachapagain) [inline docs/Developers/Design/CRD_design.md]: First sentence is fine but should be at the beginning.
Rest is not required.

**Review by umangachapagain — CHANGES_REQUESTED** [review]
Overall,
1. Deployment steps should be part of README in the root directory.
2. Purpose of `/docs/index.md` is not clear.
3. If we do not have any meaningful content for `/docs/Users_guide/index.md`, it is better we remove this for now.
4. `/docs/Developers/Design/deploy_procedure_design.md` needs to be concise.
5. `/docs/Developers/Design/CRD_design.md` needs some restructuring.

**Comment by shtripat** [comment]
> Overall,
> 
>     1. Deployment steps should be part of README in the root directory.

Ack. Will move there.

> 
>     2. Purpose of `/docs/index.md` is not clear.

Will remove this.

> 
>     3. If we do not have any meaningful content for `/docs/Users_guide/index.md`, it is better we remove this for now.

Ack.

> 
>     4. `/docs/Developers/Design/deploy_procedure_design.md` needs to be concise.

Concise in the sense? Deployment steps we are moving to README.md. Any other changes you need?

> 
>     5. `/docs/Developers/Design/CRD_design.md` needs some restructuring.

File: docs/Users_guide/index.md (hunk 1 above)
Comment (shtripat) [reply docs/Users_guide/index.md]: :) Ack. Will remove this statement. Effect of copy + paste ;)

File: docs/Developers/Design/deploy_procedure_design.md (hunk 2 above)
Comment (shtripat) [reply docs/Developers/Design/deploy_procedure_design.md:3]: Ack. Will remove.

File: docs/Developers/Design/deploy_procedure_design.md (hunk 3 above)
Comment (shtripat) [reply docs/Developers/Design/deploy_procedure_design.md]: Ack

File: docs/Developers/Design/deploy_procedure_design.md (hunk 4 above)
Comment (shtripat) [reply docs/Developers/Design/deploy_procedure_design.md]: Ack

File: docs/Developers/Design/deploy_procedure_design.md (hunk 5 above)
Comment (shtripat) [reply docs/Developers/Design/deploy_procedure_design.md]: Ok.

File: docs/Developers/Design/CRD_design.md (hunk 14 above)
Comment (shtripat) [reply docs/Developers/Design/CRD_design.md]: Ack

File: docs/Developers/Design/CRD_design.md (hunk 15 above)
Comment (shtripat) [reply docs/Developers/Design/CRD_design.md]: Ack :)

File: docs/Developers/Design/CRD_design.md (hunk 16 above)
Comment (shtripat) [reply docs/Developers/Design/CRD_design.md:78]: Got it. Will do that.

File: docs/Developers/Design/CRD_design.md (hunk 18 above)
Comment (shtripat) [reply docs/Developers/Design/CRD_design.md]: Ack

File: docs/Developers/Design/CRD_design.md (hunk 10 above)
Comment (shtripat) [reply docs/Developers/Design/CRD_design.md]: Ack

File: docs/Developers/Design/CRD_design.md (hunk 12 above)
Comment (shtripat) [reply docs/Developers/Design/CRD_design.md:30]: Ack

File: README.md [hunk 19]
```diff
@@ -1,3 +1,37 @@
...
 
 This operator takes care of deployment of mixins of storage system in Openshift and Kubernetes. The mixins are the mechanism for adding alerting rules, grafana dashboards and recording rules for prometheus in Openshift and Kubernetes.
+
+# Deploy Operator with mixins
+To deploy the operator simply deploy to your cluster. The set of steps to be followed
+for deployment of monstorak operator and related artifacts, follow the below steps
+
+## Create the namespace `storage-monitoring`
+```
+oc create -f deploy/kubernetes/01_namespace.yaml
```
Comment (cloudbehl) [inline README.md:15]: I don't see the kubernetes folder inside deploy right now.

File: README.md [hunk 20]
```diff
@@ -1,3 +1,37 @@
...
+```
+
+## Deploy the operator
+```
+oc create -f deploy/kubernetes/04_operator.yaml
+```
+
+## Deploy the custom resources
+```
+oc create -f deploy/kubernets/05_storageAlerts.yaml
```
Comment (cloudbehl) [inline README.md]: Typo : /s/kubernets/kubernetes

File: README.md [hunk 21]
```diff
@@ -1,3 +1,37 @@
...
+```
+
+## Deploy the storage alert rules
+```
+oc create -f jsonnet/manifests/ceph-prometheus-rules.yaml
```
Comment (cloudbehl) [inline README.md:20]: We are not specific to ceph to can we specify that.

File: README.md [hunk 22]
```diff
@@ -1,3 +1,37 @@
 # Monitoring operator for storages
 
 This operator takes care of deployment of mixins of storage system in Openshift and Kubernetes. The mixins are the mechanism for adding alerting rules, grafana dashboards and recording rules for prometheus in Openshift and Kubernetes.
```
Comment (cloudbehl) [inline README.md]: We can't add dashboard to openshift. So can we just specify the kubernetes.

File: README.md [hunk 23]
```diff
@@ -1,3 +1,37 @@
...
+
+# Deploy Operator with mixins
+To deploy the operator simply deploy to your cluster. The set of steps to be followed
+for deployment of monstorak operator and related artifacts, follow the below steps
```
Comment (cloudbehl) [inline README.md]: can we simplify the steps? Writing 5 different commands is a tedious task. If we can have single description and add these steps in format that a person can copy paste and done

**Review by cloudbehl — CHANGES_REQUESTED** [review]

File: README.md (hunk 19 above)
Comment (shtripat) [reply README.md:15]: Dependent on https://github.com/monstorak/monstorak/pull/24

File: README.md (hunk 20 above)
Comment (shtripat) [reply README.md]: Ack

File: README.md (hunk 21 above)
Comment (shtripat) [reply README.md:20]: Ack

File: README.md (hunk 22 above)
Comment (shtripat) [reply README.md]: Ack

File: README.md (hunk 23 above)
Comment (shtripat) [reply README.md]: Ack

**Comment by umangachapagain** [comment]
Being tracked at monstorak/monstorak/pull/31

**Pull request closed without merging**